### DISCUSSION TITLE
Custom NFT TokenMetadata for research project

### DISCUSSION THREAD

**Discussion by ghost** [near/NEPs#392] [Q&A] [answered]
Hi, I am trying to create a custom NFT with the following TokenMetadata

// Following parameters belong to the wireguard config file section [Interface]
pub address: Ipv4Addr, // IPAdress
pub listenport: u16, // Port number
pub privatekey: String, //  Ending in "="
pub postup: Option<String>,  // ip command
pub postdown: Option<String>, // ip command
pub predown: Option<String>, // ip command
pub vpnaccelerator: Option<String>, // Only on OR off
pub saveconfig: String, // Only true OR false
// Following parameters belong to the wireguard config file section [Peer]
pub publickey: String, // Ending in =
pub allowedips: Ipv4Addr, //  IPAdress range?
pub dns: Ipv4Addr, // DNS IPAdress
pub endpoint: Ipv4Addr, // IPAddress and port
 // Following parameters are part of the cableguard spec specific section [Signature]
pub authornftcontractid: String, // ID of authornftcontract
pub authorsignature: String,  //  Hash of the bbadge signed with authornftcontractid's publickey
pub bandwidthpersecond: u64, // Bandwith of the subscription in Kb/s
pub requestspersecond: u64, // Requests per second of the subscription
pub authorizedlocation: String, // From what region the subscription is valid
pub authorizednetwork: Ipv4Addr, // From what network range the subscription is valid

This is just a draft design, probably some fields can be left out of the final implementation.

I would like to store this data on-chain, in the NFT itself as it is not too heavy. The idea is to modify boringtun (a userspace implementation of wireguard in written in Rust), and create a proof of concept whereas you can purchase a VPN subscription with NEAR and upon receiving the NFT, the connection is automatically configured, using the blockchain public ed25519 keys. This has many advantages, too long to list here.

Is it possible to add any data to TokenMetadata at all, or does it require modifying NEP-177?

**Comment by willemneal** [comment, marked as answer]
I would a new field to the current metadata spec for a struct with the new data. Check out https://github.com/near/near-sdk-rs for the standarda code to add to.

**Reply by ghost** [reply]
That is what I did in the 1.skeleton branch template I was using, but the following command throws errors when I try to initialize the contract: "near call $NFT_CONTRACT_ID nft_mint '{"token_id": "token-1", "metadata": {"title": "My Non Fungible Team Token", "description": "The Team Most Certainly Goes :)", "media": "https://bafybeiftczwrtyr3k7a2k4vutd3amkwsmaqyhrdzlhvpt33dyjivufqusq.ipfs.dweb.link/goteam-gif.gif"}, "receiver_id": "'$NFT_CONTRACT_ID'"}' --accountId $NFT_CONTRACT_ID --amount 0.1", so I assume I need to make matching changes elsewhere, I need to determine where exactly.

**Reply by willemneal** [reply]
Do you have a link to your repo? Also what error are you getting?

**Reply by ghost** [reply]
@willemneal I think I am a bonehead. By changing the additional data field from "ipaddress <Ipv4Addr>" to "ipaddress Option<Ipv4Addr>", like the rest of the fields in TokenMetadata, the problem went away. I still have to run more tests, but I think that was the problem.
